security/fma-l1-upgrades.md

```diff
@@ -0,0 +1,329 @@
+4. The use of state change validation prior to execution.
+
+Additionally, the [proposal](../protocol/l1-upgrades.md#modfications-to-l1-contracts) of adding a new minimally invasive
+`upgrade()` method to L1 contracts ensures minimal storage write operations during the upgrade process.
+
+#### **Detection:**
+
+Unexpected storage layout changes which are not caught by the mitigations above will be very difficult to detect, unless they
+immediately impact on the operation of the system. For example changing the value of `L1CrossDomainMessenger.otherMessenger()`
+will break both the messenger's `relayMessage()` and `sendMessage()` functions. Other values, such as an entry in a mapping or
+a rarely used address will be harder to detect.
+
+Regardless, strong emphasis should be placed on mitigation here.
+
+#### **Recovery Path(s)**:
+
+Emergency upgrade to fix the affected storage slots.
+
+### FM3: Storage Layout Corruption in the Upgrade Controller Safe
+
+#### **Description:**
+
+The decision was made to have the Upgrade Controller Safe DELEGATECALL
+to `OPCM.upgrade()`, as this reduces the risk needing to transfer ownership to the OPCM.
+
+However this creates the risk of modifying storage in the Upgrade Controller Safe.
+
+#### **Risk Assessment:**
+
+High impact, Low likelihood
+
+#### **Mitigations:**
+
+Similar to FM2, we have existing mitigation patterns in place which we can employ here:
+
+1. The use of off-chain checks in the `superchain-ops` scripts to ensure that getters
+   continue to return the expected values.
+2. The use of state change validation prior to execution.
+
+We can also add static analysis to ensure that no `SSTORE` operations occur
+within `OPCM.upgrade()`. Slither would likely be a good approach here, or
+another tool which operates on the IR.
+
+#### **Detection:**
+
+As with FM2, unless something breaks immediately, this will be very hard to
+detect. Strong emphasis should be placed on mitigation.
+
+#### **Recovery Path(s)**:
+
+It may be impossible to recover from this, depending on which storage changes have been made.
+
+Perhaps now is the time to consider adding the Recover Module.
+
+### FM4: Failure to follow upgrade path
+
+#### **Description:**
+
+For an OP Chain upgrading though multiple releases, there is a risk that the correct upgrade path is not followed.
+
+For example, skipping from v2.0.0 to v2.2.0 without going through v2.1.0.
+
+```mermaid
+graph LR
+    A[2.0.0] --> B[2.2.0]
+
+    %% Invisible elements to help center 2.1.0
+    A ~~~C[2.1.0]
+    C ~~~B
+```
+
+This is a particular risk we have chosen to accept as opposed to enforcing the upgrade path onchain,
+because doing so comes with operational complications.
+
+#### **Risk Assessment:**
+
+Medium impact, Medium likelihood
+
+#### **Mitigations:**
+
+1. Define the correct upgrade path in the `superchain-registry`.
+2. Enforce the correct upgrade path in `op-deployer` which is the official interface for upgrading OP Chains.
```

Comment:
I wonder if it would be useful to have a solidity script in superchain-ops that helps with double checking the OPCM address is correct for the current version by using a library like: https://github.com/memester-xyz/surl

Reply:
In addition the the op-deployer check, this would add some defense in depth.

Reply:
Interesting idea. One of the main open gaps is to clarify how superchain-ops fits into the flow and where the line is between it and op-deployer.